src/Raven.Server/Documents/Queries/LuceneIntegration/TermsMatchQuery.cs

```diff
@@ -19,115 +22,191 @@ public TermsMatchQuery(string field, List<string> matches)
             Field = field;
             Matches = matches;
             Matches.Sort(StringComparer.Ordinal);
-            RewriteMethod = CONSTANT_SCORE_FILTER_REWRITE;
         }
 
-        public override string ToString(string fld)
+
+        public override Weight CreateWeight(Searcher searcher, IState state)
         {
-            return "@in<" + Field + ">(" + string.Join(", ", Matches) + ")";
+            return new TermMatchQueryWeight(this, searcher);
         }
 
-        protected override FilteredTermEnum GetEnum(IndexReader reader, IState state)
+        private class TermMatchQueryWeight : Weight
         {
-            return new RavenTermsFilteredTermEnum(this, reader, state);
-        }
+            private readonly TermsMatchQuery _parent;
+            private readonly Searcher _searcher;
+            private float _queryWeight = 1.0f;
+
+            public TermMatchQueryWeight(TermsMatchQuery parent, Searcher searcher)
+            {
+                _parent = parent;
+                _searcher = searcher;
+            }
+            public override Lucene.Net.Search.Explanation Explain(IndexReader reader, int doc, IState state)
+            {
+                var result = new ComplexExplanation {Description = _parent.ToString(), Value = _queryWeight};
+                result.AddDetail(new Lucene.Net.Search.Explanation(_queryWeight, "queryWeight"));
+                return result;
+            }
 
-        private sealed class RavenTermsFilteredTermEnum : FilteredTermEnum
+            public override void Normalize(float norm)
+            {
+                _queryWeight *= norm;
+            }
+
+            public override Scorer Scorer(IndexReader reader, bool scoreDocsInOrder, bool topScorer, IState state)
+            {
+                Similarity similarity = _parent.GetSimilarity(_searcher);
+                if(_parent.Matches.Count > 128)
+                    return new LazyInitTermMatchScorer(_parent, reader, state, similarity);
+
+                var scorers = new Scorer[_parent.Matches.Count];
```

Comment:
It would be nice to use `ArrayPool` here but I don't see where we could return it back to the pool.

Reply:
We can probably override the `DisjunctionMaxScorer` methods and detect when it is at end. Not sure how useful that would be.